File: airbyte-db/db-lib/src/test/java/io/airbyte/db/instance/configs/migrations/V0_50_7_001__AddSupportStateToActorDefinitionVersionTest.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.db.instance.configs.migrations;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.airbyte.db.factory.FlywayFactory;
import io.airbyte.db.instance.configs.AbstractConfigsDatabaseTest;
import io.airbyte.db.instance.configs.ConfigsDatabaseMigrator;
import io.airbyte.db.instance.configs.migrations.V0_50_7_001__AddSupportStateToActorDefinitionVersion.SupportState;
import io.airbyte.db.instance.development.DevDatabaseMigrator;
import java.io.IOException;
import java.sql.SQLException;
import org.flywaydb.core.Flyway;
import org.flywaydb.core.api.migration.BaseJavaMigration;
import org.jooq.DSLContext;
import org.jooq.impl.DSL;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class V0_50_7_001__AddSupportStateToActorDefinitionVersionTest extends AbstractConfigsDatabaseTest {

  @BeforeEach
  void beforeEach() {
    final Flyway flyway =
        FlywayFactory.create(dataSource, "V0_50_7_001__AddSupportStateTest.java", ConfigsDatabaseMigrator.DB_IDENTIFIER,
            ConfigsDatabaseMigrator.MIGRATION_FILE_LOCATION);
    final ConfigsDatabaseMigrator configsDbMigrator = new ConfigsDatabaseMigrator(database, flyway);

    final BaseJavaMigration previousMigration = new V0_50_6_002__AddDefaultVersionIdToActor();
    final DevDatabaseMigrator devConfigsDbMigrator = new DevDatabaseMigrator(configsDbMigrator, previousMigration.getVersion());
    devConfigsDbMigrator.createBaseline();
  }

  @Test
  void test() throws IOException, SQLException {
    final DSLContext context = getDslContext();
    V0_50_7_001__AddSupportStateToActorDefinitionVersion.addSupportStateType(context);
    V0_50_7_001__AddSupportStateToActorDefinitionVersion.addSupportStateColumnToActorDefinitionVersion(context);
    assertTrue(typeExists(context, "support_state"));
    assertTrue(columnExists(context, "support_state", "actor_definition_version"));

    // All support states should be set to "supported" after migration that added the column
    assertEquals(
        context.fetchCount(
            DSL.select()
                .from("actor_definition_version")),
        context.fetchCount(
            DSL.select()
                .from("actor_definition_version")
                .where(DSL.field("support_state").eq(SupportState.supported))));
  }

  static boolean typeExists(final DSLContext ctx, final String typeName) {
    return ctx.fetchExists(DSL.select()
        .from("pg_type")
        .where(DSL.field("typname").eq(typeName)));
  }

  static boolean columnExists(final DSLContext ctx, final String columnName, final String tableName) {
    return ctx.fetchExists(DSL.select()
        .from("information_schema.columns")
        .where(DSL.field("table_name").eq(tableName)
            .and(DSL.field("column_name").eq(columnName))));
  }

}

File: airbyte-db/db-lib/src/test/kotlin/io/airbyte/db/instance/configs/migrations/V0_50_7_001__AddSupportStateToActorDefinitionVersionTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.db.instance.configs.migrations

import io.airbyte.db.factory.FlywayFactory.create
import io.airbyte.db.instance.configs.AbstractConfigsDatabaseTest
import io.airbyte.db.instance.configs.ConfigsDatabaseMigrator
import io.airbyte.db.instance.configs.migrations.V0_50_7_001__AddSupportStateToActorDefinitionVersion.Companion.addSupportStateColumnToActorDefinitionVersion
import io.airbyte.db.instance.configs.migrations.V0_50_7_001__AddSupportStateToActorDefinitionVersion.Companion.addSupportStateType
import io.airbyte.db.instance.development.DevDatabaseMigrator
import org.flywaydb.core.api.migration.BaseJavaMigration
import org.jooq.DSLContext
import org.jooq.impl.DSL
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.io.IOException
import java.sql.SQLException

@Suppress("ktlint:standard:class-naming")
@Disabled
internal class V0_50_7_001__AddSupportStateToActorDefinitionVersionTest : AbstractConfigsDatabaseTest() {
  @BeforeEach
  fun beforeEach() {
    val flyway =
      create(
        dataSource,
        "V0_50_7_001__AddSupportStateTest.java",
        ConfigsDatabaseMigrator.DB_IDENTIFIER,
        ConfigsDatabaseMigrator.MIGRATION_FILE_LOCATION,
      )
    val configsDbMigrator = ConfigsDatabaseMigrator(database, flyway)

    val previousMigration: BaseJavaMigration = V0_50_6_002__AddDefaultVersionIdToActor()
    val devConfigsDbMigrator = DevDatabaseMigrator(configsDbMigrator, previousMigration.version)
    devConfigsDbMigrator.createBaseline()
  }

  @Test
  @Throws(IOException::class, SQLException::class)
  fun test() {
    val context = getDslContext()
    addSupportStateType(context)
    addSupportStateColumnToActorDefinitionVersion(context)
    Assertions.assertTrue(typeExists(context, "support_state"))
    Assertions.assertTrue(columnExists(context, "support_state", "actor_definition_version"))

    // All support states should be set to "supported" after migration that added the column
    Assertions.assertEquals(
      context.fetchCount(
        DSL
          .select()
          .from("actor_definition_version"),
      ),
      context.fetchCount(
        DSL
          .select()
          .from("actor_definition_version")
          .where(
            DSL
              .field("support_state")
              .eq(V0_50_7_001__AddSupportStateToActorDefinitionVersion.SupportState.supported),
          ),
      ),
    )
  }

  companion object {
    fun typeExists(
      ctx: DSLContext,
      typeName: String?,
    ): Boolean =
      ctx.fetchExists(
        DSL
          .select()
          .from("pg_type")
          .where(DSL.field("typname").eq(typeName)),
      )

    fun columnExists(
      ctx: DSLContext,
      columnName: String?,
      tableName: String?,
    ): Boolean =
      ctx.fetchExists(
        DSL
          .select()
          .from("information_schema.columns")
          .where(
            DSL
              .field("table_name")
              .eq(tableName)
              .and(DSL.field("column_name").eq(columnName)),
          ),
      )
  }
}
